File: Algorithms/120. Triangle/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
class Solution(object):
    def __init__(self):
        self.map={}
        self.tri=[[]]
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        self.tri=triangle
        self.map={}
        return self.f(0,0)
    def f(self,x,y):
        if x>=len(self.tri):
            return 0
        ans=self.map.get(str(x)+','+str(y))
        if ans is  None:
            ans=min(self.f(x+1,y),self.f(x+1,y+1))+self.tri[x][y]
            self.map[str(x)+','+str(y)]=ans
        return ans

File: Algorithms/120. Triangle/Solution.java
public class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int[] ans=new int[triangle.get(triangle.size()-1).size()];
        for (int i = 0; i <triangle.get(triangle.size()-1).size(); i++) {
            ans[i]=triangle.get(triangle.size()-1).get(i);
        }
        if (triangle.size()>1) {
            for (int i = triangle.size()-2; i >=0; i--) {
                List<Integer> list=triangle.get(i);
                for (int j = 0; j < list.size(); j++) {
                    ans[j]=list.get(j)+(ans[j]<ans[j+1]?ans[j]:ans[j+1]);
                }
            }
        }
        return ans[0];
    }
}